Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -574,88 +660,107 @@ void CGContextRestoreGState(CGContextRef context) {
-    }
+    FAIL_FAST_IF_FAILED(deviceContext->EndDraw());
+    FAIL_FAST_IF_FAILED(commandList->Close());
 
-    if (drawMode & kCGPathStroke) {
-        ComPtr<ID2D1Factory> factory;
-        deviceContext->GetFactory(&factory);
-
-        // TODO(DH): GH#1077 Do not recreate for every drawing operation
-        ComPtr<ID2D1StrokeStyle> strokeStyle;
-        // dashes must be adjusted to be based on line width
-        std::vector<CGFloat> adjustedDashes{ currentState.dashes };
-        for (float& f : adjustedDashes) {
-            f /= currentState.lineWidth;
-        }
-        FAIL_FAST_IF_FAILED(
-            factory->CreateStrokeStyle(currentState.strokeProperties, adjustedDashes.data(), adjustedDashes.size(), &strokeStyle));
+    deviceContext->SetTarget(originalTarget.Get());
 
-        deviceContext->DrawGeometry(geometry, currentState.strokeBrush.Get(), currentState.lineWidth, strokeStyle.Get());
-    }
+    *outCommandList = commandList.Detach();
+}
 
-    deviceContext->EndDraw();
-    commandList->Close();
+static void __CGContextRenderCommandList(CGContextRef context, ID2D1CommandList* commandList) {
+    ComPtr<ID2D1RenderTarget> renderTarget = context->RenderTarget();
+    auto& state = context->CurrentGState();
+
+    ComPtr<ID2D1DeviceContext> deviceContext;
+    FAIL_FAST_IF_FAILED(renderTarget.As(&deviceContext));
+
+    D2D1_SIZE_F targetSize = renderTarget->GetSize();
+    CGAffineTransform& currentTransform = state.transform;
+
+    // CG is anchored in the bottom left, but D2D is anchored in the top left.
+    // Flip the context.
+    CGAffineTransform transform = CGAffineTransformScale(currentTransform, 1.0, -1.0);
+    // Translate it back onscreen.
+    transform = CGAffineTransformTranslate(transform, 0., targetSize.height);
+    deviceContext->SetTransform(__CGAffineTransformToD2D_F(transform));
 
     deviceContext->BeginDraw();
-    deviceContext->SetTarget(originalTarget.Get());
 
     bool layer = false;
-    if (currentState.alpha != 1.0f || false /* mask, clip, etc. */) {
+    if (state.alpha != 1.0f || false /* mask, clip, etc. */) {
         layer = true;
-        renderTarget->PushLayer(D2D1::LayerParameters(D2D1::InfiniteRect(),
-                                                      nullptr,
-                                                      D2D1_ANTIALIAS_MODE_PER_PRIMITIVE,
-                                                      D2D1::IdentityMatrix(),
-                                                      currentState.alpha),
-                                nullptr);
+        renderTarget->PushLayer(
+            D2D1::LayerParameters(D2D1::InfiniteRect(), nullptr, D2D1_ANTIALIAS_MODE_PER_PRIMITIVE, D2D1::IdentityMatrix(), state.alpha),
+            nullptr);
     }
 
-    deviceContext->DrawImage(commandList.Get());
+    deviceContext->DrawImage(commandList);
 
     if (layer) {
         renderTarget->PopLayer();
     }
 
     FAIL_FAST_IF_FAILED(deviceContext->EndDraw());
```

Comment:
>FAIL_FAST_IF_FAILED [](start = 4, length = 19)

[https://msdn.microsoft.com/en-us/library/windows/desktop/dd371766(v=vs.85).aspx](https://msdn.microsoft.com/en-us/library/windows/desktop/dd371766(v=vs.85).aspx), we should not failfast if the EndDraw command results in D2DERR_RECREATE_TARGET.  It can happen if the graphics driver stops responding, and the only way is to recreate the target and replay the commands.